File: jupyter2/problems/recursion/fast_power/ml/fast_power.ml
let rec power_iter a n =
  let p = ref 1.0 in
  (for i = 1 to n do
     p := !p *. a
   done;
   !p)
(** begin my answer *)

(** end my answer *)

let main () =
  assert (Float.abs (power_simple (1.0 +. 1.0 /. 10.0)      10        -. 2.59374246) < 1.0e-7);
  assert (Float.abs (power_fast (1.0 +. 1.0 /. 10.0)        10        -. 2.59374246) < 1.0e-7);
  assert (Float.abs (power_simple (1.0 +. 1.0 /. 100.0)     100       -. 2.70481383) < 1.0e-7);
  assert (Float.abs (power_fast (1.0 +. 1.0 /. 100.0)       100       -. 2.70481383) < 1.0e-7);
  assert (Float.abs (power_fast (1.0 +. 1.0 /. 1000000.0)   1000000   -. 2.71828047) < 1.0e-7);
  assert (Float.abs (power_iter (1.0 +. 1.0 /. 1000000.0)   1000000   -. 2.71828047) < 1.0e-7);
  assert (Float.abs (power_fast (1.0 +. 1.0 /. 10000000.0)  10000000  -. 2.71828169) < 1.0e-7);
  assert (Float.abs (power_iter (1.0 +. 1.0 /. 10000000.0)  10000000  -. 2.71828169) < 1.0e-7);
  assert (Float.abs (power_fast (1.0 +. 1.0 /. 100000000.0) 100000000 -. 2.71828179) < 1.0e-7);
  assert (Float.abs (power_iter (1.0 +. 1.0 /. 100000000.0) 100000000 -. 2.71828180) < 1.0e-7);
  Printf.printf "OK\n"
;;

main()

File: jupyter2/problems/recursion/fast_power/rs/fast_power.rs
fn power_iter(a: f64, n: i64) -> f64 {
    let mut p = 1.0;
    for _i in 1..n+1 {
        p *= a
    }
    p
}
/** begin my answer */

/** end my answer */

fn main() {
    assert!((power_slow(1.0 + 1.0/10.0,        10).abs()       - 2.59374246) < 1.0e-7);
    assert!((power_fast(1.0 + 1.0/10.0,        10).abs()       - 2.59374246) < 1.0e-7);
    assert!((power_slow(1.0 + 1.0/100.0,       100).abs()      - 2.70481383) < 1.0e-7);
    assert!((power_fast(1.0 + 1.0/100.0,       100).abs()      - 2.70481383) < 1.0e-7);
    assert!((power_fast(1.0 + 1.0/1000000.0,   1000000).abs()  - 2.71828047) < 1.0e-7);
    assert!((power_iter(1.0 + 1.0/1000000.0,   1000000).abs()  - 2.71828047) < 1.0e-7);
    assert!((power_fast(1.0 + 1.0/10000000.0,  10000000).abs() - 2.71828169) < 1.0e-7);
    assert!((power_iter(1.0 + 1.0/10000000.0,  10000000).abs() - 2.71828169) < 1.0e-7);
    assert!((power_fast(1.0 + 1.0/100000000.0, 100000000).abs() - 2.71828179) < 1.0e-7);
    assert!((power_iter(1.0 + 1.0/100000000.0, 100000000).abs() - 2.71828180) < 1.0e-7);
    println!("OK");
}
